src/validation.cpp

```diff
@@ -3410,10 +3419,21 @@ bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<cons
 
         LOCK(cs_main);
 
-        if (ret) {
+        if (!ret) {
+            // Mark the block as invalid if we recognize it in mapBlockIndex.
+            // This doesn't happen within CheckBlock so we have to include a call
+            // here. It *does* happen within AcceptBlock below.
+            BlockMap::iterator it = mapBlockIndex.find(pblock->GetHash());
+
+            // Duplicate call to CorruptionPossible() here just to be careful.
```

Comment:
In commit "Do the proper InvalidateBlock bookkeeping for DoSy blocks"

Duplicating the CorruptionPossible check here seems unnecessarily confusing and fragile to me, though perhaps there's an advantage I'm not seeing.